filtering.py

```diff
@@ -0,0 +1,941 @@
+        else:
+            self.target_models = None
+
+    def forward(self, audio, rate):
+        """
+        Performing the separation on audio input
+
+        Parameters
+        ----------
+        audio: torch.Tensor [shape=(nb_samples, nb_channels, nb_timesteps)]
+        mixture audio
+
+        rate: int
+        sampling rate of the audio.
+
+        Returns
+        -------
+        estimates: `dict` [`str`, `torch.Tensor`]
+            dictionary of all restimates as performed by the separation model.
+            Note that there may be an additional source in case
+            a residual source is added. If the input audio is 2D, then the
+            first dimension (sample) is removed also from the input.
+
+            Note however, that the last two dimensions are
+            (channel, time steps) in all cases.
+        model_rate: int
+            the new sampling rate, desired by the open-unmix model
+            """
+        if not utils._torchaudio_available():
+            raise Exception('Torchaudio must be available to use the'
+                            'Separator.')
+        import torchaudio
+        from torchaudio.functional import istft
+
+        # shape management
+        remove_samples_dim = False
+        convert_to_numpy = False
+        if self.smart_input_management:
+            shape = torch.tensor(audio.shape)
+            if not isinstance(audio, torch.Tensor):
+                audio = torch.tensor(audio, device=self.device,
+                                     requires_grad=False)
+                convert_to_numpy = True
+            if len(shape) == 1:
+                audio = audio[None, None, ...]
+                remove_samples_dim = True
+            elif len(shape) == 2:
+                if shape.min() <= 2:
+                    audio = audio[None, ...]
+                    remove_samples_dim = True
+                else:
+                    audio = audio[:, None, ...]
+            if audio.shape[1] > audio.shape[2]:
+                audio = audio.transpose(1, 2)
+            if audio.shape[1] > 2:
+                warnings.warn(
+                    'Channel count > 2!. Only the first two channels '
+                    'will be processed!')
+                audio = audio[..., :2]
+        audio = audio.float()
+        if audio.shape[1] == 1:
+            # if we have mono, we duplicate it
+            # as the input of OpenUnmix is always stereo
+            audio = audio.expand(-1, 2, -1)
+
+        V = None
+
+        nb_sources = len(self.targets)
+        nb_samples = audio.shape[0]
+
+        for j, target in enumerate(self.targets):
+            # if the models have not been pre-loaded, load it now
+            if not self.training:
+                unmix_target = model.load_model(
+                    target=target,
+                    model_name=self.model_name,
+                    device=self.device
+                )
+                unmix_target.freeze()
```

Comment:
i think we should not do his is forward path. I know you did this to prevent high ram usage, but lets not care about this and think about training as well (where all models would need to be kept into into memory). So lets put this in init